vpr/src/route/rr_graph_fpga_interchange.cpp

```diff
@@ -0,0 +1,1179 @@
+        links[0] = links[1] = links[2] = links[3] = nullptr;
+        visited = nullptr;
+        on_route = false;
+        has_pins = false;
+    }
+};
+
+/*
+ * Main class for FPGA Interchange device to rr_graph flow
+ */
+struct RR_Graph_Builder {
+  public:
+    RR_Graph_Builder(Device::Reader& arch_reader,
+                     const DeviceGrid& grid,
+                     DeviceContext& device_ctx,
+                     const std::vector<t_physical_tile_type>& physical_tile_types,
+                     const std::vector<t_segment_inf>& segment_inf,
+                     const enum e_base_cost_type base_cost_type)
+        : ar_(arch_reader)
+        , device_ctx_(device_ctx)
+        , segment_inf_(segment_inf)
+        , grid_(grid)
+        , base_cost_type_(base_cost_type)
+        , empty_(device_ctx.arch->strings.intern_string(vtr::string_view(""))) {
+        for (auto& type : physical_tile_types) {
+            tile_type_to_pb_type_[std::string(type.name)] = &type;
+        }
+
+        for (size_t i = 0; i < segment_inf.size(); ++i) {
+            wire_name_to_seg_idx_[segment_inf[i].name] = i;
+        }
+
+        for (size_t i = 1; i < grid.width() - 1; ++i) {
+            for (size_t j = 1; j < grid.height() - 1; ++j) {
+                location loc{i, j};
+                if (grid[i][j].type->index == 0)
+                    empty_tiles.insert(loc);
+            }
+        }
+    }
+
+    void build_rr_graph() {
+        create_switch_inf();
+        create_tile_id_to_loc();
+        create_uniq_node_ids();
+        create_sink_source_nodes();
+        create_pips_();
+        process_nodes();
+        pack_to_rr_graph();
+#ifdef DEBUG
+        print_();
+#endif
+        finish_rr_generation();
+    }
+
+  private:
+    Device::Reader& ar_;
+    DeviceContext& device_ctx_;
+    const std::vector<t_segment_inf>& segment_inf_;
+    const DeviceGrid& grid_;
+    const enum e_base_cost_type base_cost_type_;
+
+    vtr::interned_string empty_;
+
+    std::unordered_map<std::string, const t_physical_tile_type*> tile_type_to_pb_type_;
+
+    std::unordered_map<std::string, int> wire_name_to_seg_idx_;
+
+    std::set<location> empty_tiles;
+
+    std::unordered_map<int /*tile_id(name)*/, location> tile_to_loc_;
+    std::map<location, int /*tile_id(name)*/> loc_to_tile_;
+
+    std::map<std::tuple<int /*timing_model*/, bool /*buffered*/>, int /*idx*/> pips_models_;
+    std::map<std::tuple<int /*node_id*/, int /*node_id*/>, std::tuple<int /*switch_id*/, std::tuple<int /*tile_id*/, int /*wire0*/, int /*wire1*/, bool /*forward*/>>> pips_;
+
+    std::map<std::tuple<int, int>, int> wire_to_node_;
+    std::unordered_map<int, std::set<location>> node_to_locs_;
+    std::map<std::tuple<int, int> /*<node, tile_id>*/, int /*segment type*/> node_tile_to_segment_;
+
+    /*
+     * Offsets for FPGA Interchange node processing
+     */
+    location offsets[4] = {location(-1, 0), location(0, 1), location(1, 0), location(0, -1)};
```

Comment:
For clarity it might be good to explicitly have a map or vector (with indices using the above mentioned enum)